### PR TITLE
favorite: saga-ize add/ignore + generating promise helpers in rpc layer

### PR DESCRIPTION
**Pull request by chrisnojima**

### CODE REVIEW

**Comment by chrisnojima** [comment]
@keybase/react-hackers this is ready to look at

File: protocol/bin/flow.js [hunk 1]
```diff
@@ -211,7 +211,11 @@ function analyzeMessages (json, project) {
     const rpc = isUIProtocol ? '' : `export function ${name}Rpc (request: $Exact<${['requestCommon', callbackType, innerParamType].filter(t => t).join(' & ')}>) {
   engineRpcOutgoing({...request, method: '${json.protocol}.${m}'})
 }`
-    return [paramType, response, rpc]
+
+    const rpcPromise = isUIProtocol ? '' : `export function ${name}RpcPromise (request: $Exact<${['requestCommon', callbackType, innerParamType].filter(t => t).join(' & ')}>): Promise<${responseType !== 'null' ? `${name}Result` : 'any'}> {
```
Comment (chrisnojima) [inline protocol/bin/flow.js]: we add new promise helpers so we can use yield call() easily

File: shared/actions/favorite.js [hunk 2]
```diff
@@ -24,71 +24,71 @@ function toggleShowIgnored (isPrivate: boolean): FavoriteToggleIgnored {
...
+
 const injectMeta = type => f => { f.meta = type }
 
 const _jsonToFolders = (json: Object, myKID: any) => {
   const folderSets = [json.favorites, json.ignored, json.new]
+  const fillFolder = folder => {
+    folder.waitingForParticipantUnlock = []
+    folder.youCanUnlock = []
 
-  folderSets.forEach(folders => {
```
Comment (chrisnojima) [inline shared/actions/favorite.js]: this part of the diff looks kinda crappy in github but its just hoisting the closure into a helper

File: shared/actions/favorite.js [hunk 3]
```diff
@@ -208,22 +196,66 @@ function _getFavoritesRPCToFolders (txt: string, username: string = '', loggedIn
...
+    yield put(action)
+    return
+  } else {
+    try {
+      yield call(favoriteFavoriteAddRpcPromise, {param: {folder}})
+      const action: FavoriteAdded = {type: Constants.favoriteAdded, payload: undefined}
+      yield put(action)
+      yield put(navigateBack())
+    } catch (error) {
+      console.warn('Err in favorite.favoriteAdd', error)
```
Comment (MarcoPolo) [inline shared/actions/favorite.js]: Shouldn't you put a `FavirteAdded` Action?

File: shared/actions/favorite.js [hunk 4]
```diff
@@ -258,59 +290,11 @@ function _notify (state) {
...
-      },
-    })
-  }
-}
-
 function * _favoriteSaga (): SagaGenerator<any, any> {
   yield [
-    takeLatest(Constants.favoriteGet, getFavoritesListSaga),
+    takeLatest(Constants.favoriteList, _listSaga),
+    takeLatest(Constants.favoriteAdd, _addSaga),
```
Comment (MarcoPolo) [inline shared/actions/favorite.js]: Should this and ignore be `takeEvery`?

**Comment by MarcoPolo** [comment]
👍 % C

File: shared/actions/favorite.js (hunk 3 above)
Comment (chrisnojima) [inline shared/actions/favorite.js]: but the thing failed. also nothing actually listens for this

**Pull request merged**